Mathlib/Data/Multiset/Order.lean

```diff
@@ -0,0 +1,513 @@
+  · have : N.count x ≥ Q.count x := by linarith
+    simp_all only [ge_iff_le, min_eq_right]
+    apply le_eq_sub (M.count x) (N.count x) (P.count x) (Q.count x)
+    · exact l_u
+    · exact H0
+  · simp_all only [not_le, gt_iff_lt]
+    have : Multiset.count x N ≤ Multiset.count x Q := by linarith
+    have := le_of_lt l_u
+    simp_all
+
+lemma in_notin_diff {α} [DecidableEq α]:
+    ∀ (x : α) (X Y: Multiset α) ,  x ∈ X → x ∉ Y → x ∈ X - Y  := by
+  intros x X Y x_in_X x_notin_Y
+  have : Multiset.count x X ≥ 1 := by
+    rw [← Multiset.one_le_count_iff_mem] at x_in_X
+    exact x_in_X
+  rw [← Multiset.one_le_count_iff_mem, Multiset.count_sub]
+  simp_all only [not_false_eq_true, Multiset.count_eq_zero_of_not_mem, tsub_zero]
+
+-- `MultisetLT` is transitive. Two lemmas needed: double_split, in_notin_diff
+lemma LT_trans {α} [pre : Preorder α] [dec : DecidableEq α]:
+    ∀ (M N P : Multiset α) , MultisetLT N M → MultisetLT P N → MultisetLT P M := by
+  intros M N P LTNM LTPN
+  rcases LTNM with ⟨Y1, X1, Z1, X1_ne, N1_def, M1_def, Ord1⟩
+  rcases LTPN with ⟨Y2, X2, Z2, _, P2_def, N2_def, Ord2⟩
+  apply MultisetLT.MLT (Y2 + (Y1 - X2)) (X1 + (X2 - Y1)) (Z1 ∩ Z2)
+  · simp_all only [Multiset.empty_eq_zero, ne_eq, add_eq_zero_iff, false_and, not_false_eq_true]
+  · rw [P2_def]
+    have : Z1 ∩ Z2 + (Y2 + (Y1 - X2)) = Z1 ∩ Z2 + (Y1 - X2) + Y2 := by
+      have : (Y2 + (Y1 - X2)) = (Y1 - X2) + Y2 := by rw [add_comm]
+      rw [this, add_assoc]
+    rw [this]
+    have : Z1 ∩ Z2 + (Y1 - X2) = Z2 ∩ Z1 + (Y1 - X2) := by
+      rw [Multiset.inter_comm]
+    rw [this,← double_split]
+    -- Unable to merge into one line. Why?
+    rw [add_comm, ← N2_def, N1_def]
+    apply add_comm
+  · rw [M1_def]
+    have : Z1 ∩ Z2 + (X1 + (X2 - Y1)) = Z1 ∩ Z2 + (X2 - Y1) + X1 := by
+      have : (X1 + (X2 - Y1)) = (X2 - Y1) + X1 := by rw [add_comm]
+      rw [this, add_assoc]
+    rw [this, add_left_inj]
+    apply double_split
+    rw [add_comm, ← N1_def, N2_def]
+    apply add_comm
+  · intros y y_in_union
+    by_cases y_in : y ∈ Y2
+    · rcases (Ord2 y y_in) with ⟨x, x_in_X2, y_lt_x⟩
+      by_cases x_in : x ∈ Y1
+      · rcases (Ord1 x x_in) with ⟨x', x'_in_X1, x_lt_x'⟩
+        use x'
+        constructor
+        · rw [Multiset.mem_add]
+          constructor
+          exact x'_in_X1
+        · exact lt_trans y_lt_x x_lt_x'
+      · use x
+        constructor
+        · rw [add_comm, Multiset.mem_add]
+          constructor
+          apply in_notin_diff
+          exact x_in_X2
+          exact x_in
+        · exact y_lt_x
+    · have y_in : y ∈ (Y1 - X2) := by simp_all only [Multiset.mem_add, false_or]
+      let h := (Ord1 y)
+      have y_in_Y1 : y ∈ Y1 := by
+        have : Y1 - X2 ≤ Y1 := by simp_all only [tsub_le_iff_right, le_add_iff_nonneg_right,
+        zero_le]
+        apply Multiset.mem_of_le
+        exact this
+        exact y_in
+      let _ := h y_in_Y1
+      aesop
+
+lemma direct_subset_red [dec : DecidableEq α] [Preorder α]
```

Comment:
The names all need to change, but this lemma needs to be named `X_of_Y` for the appropriate X and Y.